Go detour pathfinder

arl opened the pull request:
Add new pathfinder with go-detour lib
Tile nav meshes
Objects (tables, etc) described in JSON are added as convex volumes.

For now the navmesh is static, meaning that it's created at server launch, correctly with the convex volumes representing the tables, sofa, etc. But that"s all, for now there aren't any in-game event that will trigger an addition or removal of those convex volumes, nor rebuilding of the navmesh. I'll do that when i come back from hollydays, but it will be the subject of another PR

arl committed:
server/vendor: add go-detour

arl committed:
server/surviveler: check for existence of walls+floor mesh in assets folder

arl committed:
server/vendor: add go-detour/sample

arl committed:
server/vendor: add go-detour/sample/solomesh

arl committed:
server: bump math32 vendor

arl committed:
server/resource: remove AssetsFolder.FullPath()

arl committed:
server/vendor: bump go-detour

arl committed:
server/resource: fix comment

arl committed:
server/surviveler: NewWorld loads nav mesh

arl committed:
server/surviveler: use navmesh to validate entities spawn points

arl committed:
server/surviveler: NewWorld also loads navmesh settings from yaml

arl committed:
client/loaders: fix decorator resource_handler

The check only only took into account when ext represented a single
file extension string. In case `ext` was more than one extension,
ResourceHandlerAlreadyExists would never have been raised.

arl committed:
client/loaders: register dummy yaml loader

arl committed:
server/surviveler: comment out World.Attach/DetachEntity

arl committed:
server/surviveler: add a NavMeshQuery to the World

arl committed:
server/surviveler: remove arg name from runPathFinder

arl committed:
server/surviveler: add TODO in movable

arl committed:
server/surviveler: add the path in reverse order in movable.SetPath

arl committed:
server/surviveler: add navmesh query filter in World

arl committed:
server/surviveler: World.PointInBounds uses navmesh

arl committed:
server/surviveler: remove World.AABBSpatialQuery, it wont be needed

arl committed:
server/surviveler: use navmesh for pathfinding

arl committed:
server/vendor: bump gogeo

arl committed:
server/math: remove files in math pkg, gogeo is used now

arl committed:
server/surviveler: use d2.Vec2 in tnrequests

arl committed:
server/vendor: bump go-detour (tile navmesh :D)

arl committed:
server/surviveler: World uses TileMesh

arl committed:
client: remove Y offsets in Map, MapObject and Terrain classes

arl committed:
client: remove 2nd Y offset MapObject. Remove FIXMEs

arl committed:
server: bump go-detour (sliced-pathfinding)

arl committed:
WIP modif pathfinder closestPointOnPolyBoundary

arl committed:
server/vendor: bump go-detour (convex volumes)

arl committed:
server/surviveler: add new JSON Object fields (BoundingBox & Note)

arl committed:
server/surviveler: setup global navmesh QueryFilter

arl committed:
server/surviveler: add convex volumes at startup

Objects are loaded from JSON (they need a new bounding_box field),
positioned and rotated, then added as convex volumes to the
navmesh before the first navmesh building.

arl committed:
tools/wallmap: cli output cosmetics

arl committed:
server/vendor: bump gogeo (2d rotations)

arl committed:
server/vendor: bump math32 (trigonometric tables)

arl committed:
server/surviveler: bounding_box_2d is pair of vec2d

arl committed:
server/surviveler: add objects as convex volumes to the navmesh

arl committed:
server/surviveler: refine World.PointInBounds

arl committed:
server/surviveler: use PointInBounds for FindPath

duploduplo commented on src/client/game/entities/map.py:
```diff
@@ -35,9 +35,7 @@ def __init__(self, resource, scene):
-        # FIXME: this offset here is due to the calculation of the walkable
-        # matrix that adds one more walkable line on top of the scenario.
-        self.obj.position = Vec(0.0, 0.0, 1.0)
+        self.obj.position = Vec(0.0, 0.0, 0.0)
```
Finally!

duploduplo commented on src/client/game/entities/map_object.py:
```diff
@@ -80,9 +78,7 @@ def update(self, dt):
-        # FIXME: this offset here is due to the calculation of the walkable
-        # matrix that adds one more walkable line on top of the scenario.
-        return l + Vec(pos[0], 0, pos[1] + 1), m + Vec(pos[0], 0, pos[1] + 1)
+        return l + Vec(pos[0], 0, pos[1]), m + Vec(pos[0], 0, pos[1])
```
FINALLY!

duploduplo commented on src/server/surviveler/assets_load.go:
```diff
@@ -7,14 +7,12 @@ package surviveler
-
 	log "github.com/Sirupsen/logrus"
+	"github.com/aurelien-rainone/go-detour/detour"
+	"github.com/aurelien-rainone/gogeo/f32/d3"
```
I would highly recommend to have those projects in Rookie GameDevs organization. C'mon dude, let's share the glory!

duploduplo commented on src/server/surviveler/assets_load.go:
```diff
@@ -128,19 +104,26 @@ func newGameData(pkg resource.Package) (*gameData, error) {
-				spawnPoints.Players[i])
-		}
-		if pt.Kind&KindWalkable == 0 {
+		fmt.Println("validating player spawn point", i, "=", spawnPoints.Players[i])
```
Is this print line needed? Shouldn't we use the log?

duploduplo commented on src/server/surviveler/assets_types.go:
```diff
@@ -23,9 +23,12 @@ type (
+	Pos      d2.Vec2 `json:"pos"`            // position of the object on the map
+	Rotation int     `json:"rotation"`       // rotation of the object on the map
+	Note     string  `json:"note,omitEmpty"` // note on the object
+	// TODO: remove the omitEmpty
```
Is this TODO relevant?

duploduplo commented on src/server/surviveler/movable.go:
```diff
@@ -126,12 +126,16 @@ func (me *Movable) Move(dt time.Duration) (hasMoved bool) {
  * It replaces and cancel the current path, if any.
  */
 func (me *Movable) SetPath(path Path) {
+	// TODO: the stack implementation should let the possibility to
```
Same here: are you going to take care of this in this pr?

duploduplo commented on src/server/surviveler/world.go:
```diff
@@ -180,26 +344,26 @@ func (w World) IntersectingTiles(bb d2.Rectangle) []*Tile {
  */
 func (w *World) AttachEntity(ent Entity) {
 	// retrieve list of tiles intersecting with the entity aabb
```
What is happening in this function?

duploduplo commented on src/server/surviveler/world.go:
```diff
@@ -180,26 +344,26 @@ func (w World) IntersectingTiles(bb d2.Rectangle) []*Tile {
  */
 func (w *World) DetachEntity(ent Entity) {
 	// retrieve tile list for this entity
-	tileList := w.Entities[ent.Id()]
```
Same here: should we remove those two functions? Is there any point in keeping them here?

duploduplo reviewed (changes requested):
I'm not proficient enough in go to actually give a full technical review. It's a huge work, and playing a bit with it I found it working just fine.

I added a few comments that I would like to be addressed before merging.

There are also a few tables (in the ex-enquos room, for example) that the character is just able to trespass: most likely we need to update the resource files to take care of those too.

BTW, it's great! Seeing the character smoothly wandering around got me tears!

arl commented:
Guys. Hi from Borneo, Malaysia! Weather is 35 degrees, water is good lol!
BTW I don't have access to a computer, it's just me and my phone!

More seriously, the TODO's that I let in the code plus some other quirks
are artifacts of the incompleteness of the new pathfinder integration and
show that there is still some work to be done, surviveler-side, as the
porting is finished.

I wanted, for time reasons, but also for matters of readability and
reviewability, to split this PR into smaller pieces. For example, I removed
the underlying grid on which entities were attached and detached: the
squares were acting as unit for the grid pathfinding, tiles for building
locations, as well as serving as a basic collision detection framework.

There is no grid anymore and it will be much more efficient to use
https://github.com/aurelien-rainone/golq.

I suggest that you merge this PR as I will adress the points evoked in it
when I get back. But if you prefer to let it as it is and merge only the
whole thing, it's up to you, no problem for me.

About the traversed tables, you could set up the Debug log level and check,
at server launch, if and how the tables in the json are interpreted and
converted to convex volumes. Debug log level should log the 4 vertices of
the ground plane.

Some other nice improvements are on their way. Before leaving I ported also
the steering behaviours library but didn't have the time to check it in
surviveler...

See you guys. I come back the 17th.


Le 4 mai 2017 02:05, "Lorenzo Berni" <notifications@github.com> a écrit :

*@duploduplo* requested changes on this pull request.

I'm not proficient enough in go to actually give a full technical review.
It's a huge work, and playing a bit with it I found it working just fine.

I added a few comments that I would like to be addressed before merging.

There are also a few tables (in the ex-enquos room, for example) that the
character is just able to trespass: most likely we need to update the
resource files to take care of those too.

BTW, it's great! Seeing the character smoothly wandering around got me
tears!
------------------------------

In src/client/game/entities/map.py
<https://github.com/RookieGameDevs/surviveler/pull/26#discussion_r114608061>
:

> @@ -35,9 +35,7 @@ def __init__(self, resource, scene):
         props.cast_shadows = False

         self.obj = scene.add_mesh(mesh, props)
-        # FIXME: this offset here is due to the calculation of the walkable
-        # matrix that adds one more walkable line on top of the scenario.
-        self.obj.position = Vec(0.0, 0.0, 1.0)
+        self.obj.position = Vec(0.0, 0.0, 0.0)

Finally!
------------------------------

In src/client/game/entities/map_object.py
<https://github.com/RookieGameDevs/surviveler/pull/26#discussion_r114608196>
:

> @@ -80,9 +78,7 @@ def update(self, dt):
     def bounding_box(self):
         l, m = self._bounding_box
         pos = self.position
-        # FIXME: this offset here is due to the calculation of the walkable
-        # matrix that adds one more walkable line on top of the scenario.
-        return l + Vec(pos[0], 0, pos[1] + 1), m + Vec(pos[0], 0, pos[1] + 1)
+        return l + Vec(pos[0], 0, pos[1]), m + Vec(pos[0], 0, pos[1])

FINALLY!
------------------------------

In src/server/surviveler/assets_load.go
<https://github.com/RookieGameDevs/surviveler/pull/26#discussion_r114608720>
:

>  	log "github.com/Sirupsen/logrus"
+	"github.com/aurelien-rainone/go-detour/detour"
+	"github.com/aurelien-rainone/gogeo/f32/d3"

I would highly recommend to have those projects in Rookie GameDevs
organization. C'mon dude, let's share the glory!
------------------------------

In src/server/surviveler/assets_load.go
<https://github.com/RookieGameDevs/surviveler/pull/26#discussion_r114609242>
:

>  	// validate player spawn point
 	spawnPoints := gd.mapData.AIKeypoints.Spawn
 	for i := range spawnPoints.Players {
-		pt := world.TileFromWorldVec(spawnPoints.Players[i])
-		if pt == nil {
-			return fmt.Errorf(
-				"player spawn point is out of bounds (%#v)",
-				spawnPoints.Players[i])
-		}
-		if pt.Kind&KindWalkable == 0 {
+		fmt.Println("validating player spawn point", i, "=", spawnPoints.Players[i])

Is this print line needed? Shouldn't we use the log?
------------------------------

In src/server/surviveler/assets_types.go
<https://github.com/RookieGameDevs/surviveler/pull/26#discussion_r114609581>
:

> @@ -23,9 +23,12 @@ type (
 )

 type MapObject struct {
-	Ref      string  `json:"ref"`      // uri of the object inside the package
-	Pos      d2.Vec2 `json:"pos"`      // position of the object on the map
-	Rotation int     `json:"rotation"` // rotation of the object on the map
+	Ref      string  `json:"ref"`            // uri of the object inside
the package
+	Pos      d2.Vec2 `json:"pos"`            // position of the object on the map
+	Rotation int     `json:"rotation"`       // rotation of the object on the map
+	Note     string  `json:"note,omitEmpty"` // note on the object
+	// TODO: remove the omitEmpty

Is this TODO relevant?
------------------------------

In src/server/surviveler/movable.go
<https://github.com/RookieGameDevs/surviveler/pull/26#discussion_r114609839>
:

> @@ -126,12 +126,16 @@ func (me *Movable) Move(dt time.Duration) (hasMoved bool) {
  * It replaces and cancel the current path, if any.
  */
 func (me *Movable) SetPath(path Path) {
+	// TODO: the stack implementation should let the possibility to

Same here: are you going to take care of this in this pr?
------------------------------

In src/server/surviveler/world.go
<https://github.com/RookieGameDevs/surviveler/pull/26#discussion_r114610262>
:

> @@ -180,26 +344,26 @@ func (w World) IntersectingTiles(bb d2.Rectangle) []*Tile {
  */
 func (w *World) AttachEntity(ent Entity) {
 	// retrieve list of tiles intersecting with the entity aabb

What is happening in this function?
------------------------------

In src/server/surviveler/world.go
<https://github.com/RookieGameDevs/surviveler/pull/26#discussion_r114610660>
:

>  }

 /*
  * DetachEntity detaches an entity from the underlying world representation
  */
 func (w *World) DetachEntity(ent Entity) {
 	// retrieve tile list for this entity
-	tileList := w.Entities[ent.Id()]

Same here: should we remove those two functions? Is there any point in
keeping them here?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/RookieGameDevs/surviveler/pull/26#pullrequestreview-36089323>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAdF6qi_vM1fVU4uU4DEeE8DLuPNCws3ks5r2MHPgaJpZM4M-YGN>
.
